nemo/collections/multimodal/data/clip/mbeir_dataset.py

```diff
@@ -0,0 +1,487 @@
+        # neg_cand_modality could be different from pos_cand_modality.
+        query_prompt = self._get_random_query_prompt(query_dataset_id, query_modality, pos_cand_modality)
+        query_txt_with_prompt = format_string(f"{query_prompt} {query_txt}")
+        query_txt_without_prompt = format_string(query_txt)
+
+        # Sample negative examples
+        selected_neg_cand_list = []
+        if self.mode == Mode.TRAIN:
+            neg_cand_id_list = mbeir_entry.get("neg_cand_list", [])
+            if self.hard_neg_num > 0:
+                assert len(neg_cand_id_list) > 0, f"Cannot find negative candidates for {mbeir_entry}"
+                if self.shuffle_cand:
+                    random.shuffle(neg_cand_id_list)
+                selected_neg_cand_id_list = []
+                for i in range(self.hard_neg_num):
+                    selected_neg_cand_id_list.append(
+                        neg_cand_id_list[i % len(neg_cand_id_list)]
+                    )  # % Wrap around from idx 0.
+                for neg_cand_did in selected_neg_cand_id_list:
+                    neg_cand = self.cand_pool.get(neg_cand_did, None)
+                    neg_cand_txt = neg_cand.get("txt") or ""
+                    neg_cand_txt = format_string(neg_cand_txt)
+                    neg_cand["txt"] = neg_cand_txt
+                    selected_neg_cand_list.append(neg_cand)
+
+        def _prepare_data_dict(txt, img_path):
+            img = self._load_and_preprocess_image(img_path)
+#             img = img_path
+            return {"txt": txt, "img": img}
+
+        query = _prepare_data_dict(
+            query_txt_with_prompt if self.enable_query_instruct else query_txt_without_prompt, query_img_path
+        )
+        instance = {"query": query}
+
+        if self.mode == Mode.EVAL:
+            if self.returns.get("hashed_qid"):
+                instance.update({"qid": hash_qid(qid)})
+            if self.returns.get("task_id"):
+                instance.update({"task_id": get_mbeir_task_id(query_modality, pos_cand_modality)})
+            # TODO: add src_content if needed
+
+        if self.mode == Mode.TRAIN:
+            if self.returns.get("hashed_p_did"):
+                instance.update({"p_did": hash_did(selected_pos_cand_did)})
+
+            pos_cand = _prepare_data_dict(
+                pos_cand_txt,
+                pos_cand.get("img_path", None),
+            )
+            instance.update({"pos_cand": pos_cand})
+
+            neg_cand_list = [
+                _prepare_data_dict(
+                    neg_cand["txt"],
+                    neg_cand.get("img_path", None),
+                )
+                for neg_cand in selected_neg_cand_list
+            ]
+            if len(neg_cand_list) > 0:
+                instance.update({"neg_cand_list": neg_cand_list})
+        return instance
+
+
+class MBEIRCandidatePoolDataset(MBEIRDatasetBase):
```

Comment:
If these dataset are fully/partially copied from an external repo, we have to make sure about their license

Reply:
This is from [uniir](https://github.com/TIGER-AI-Lab/UniIR/tree/main/src/data), its MIT license. I can add the link in the comments in the file.